onnxruntime/core/optimizer/computation_reduction.cc

```diff
@@ -0,0 +1,218 @@
+  Note: b: batch, s: sequence_length, h: hidden_size, p_s: dynamic_prediction_count
+*/
+static Status MatMulHandler(Graph& graph, Node* gathernd_node, Node* gathernd_input_node) {
+  int non_weight_input_index = -1;
+  for (size_t i = 0; i < gathernd_input_node->MutableInputDefs().size(); i++) {
+    if (!graph_utils::IsGraphInput(graph, gathernd_input_node->MutableInputDefs()[i]) &&
+        !graph_utils::IsInitializer(graph, gathernd_input_node->MutableInputDefs()[i]->Name(), false)) {
+      non_weight_input_index = i;
+    }
+  }
+
+  ORT_RETURN_IF_NOT(non_weight_input_index == 0);
+  auto output_shape = gathernd_input_node->MutableOutputDefs()[0]->Shape();
+  auto input_shape = gathernd_input_node->MutableInputDefs()[non_weight_input_index]->Shape();
+
+  ORT_RETURN_IF_NOT(output_shape != nullptr && input_shape != nullptr);
+  const int output_rank = output_shape->dim_size();
+  const int input_rank = input_shape->dim_size();
+  ORT_RETURN_IF_NOT(input_rank == output_rank);
+  // Only hande cases where input share exactly equal with output shape on the first two dims (
+  // because GatherND's batch_dims be 1).
+  // otherwise, it might not be safe to move GatherND upward.
+  for (int i = 0; i < 2; i++) {
+    auto& output_dim = output_shape->dim(i);
+    auto& input_dim = input_shape->dim(i);
+    if (output_dim.has_dim_value() && input_dim.has_dim_value()) {
+      ORT_RETURN_IF_NOT(output_dim.dim_value() == input_dim.dim_value());
+    } else if (output_dim.has_dim_param() && input_dim.has_dim_param()) {
+      ORT_RETURN_IF_NOT(output_dim.dim_param() == input_dim.dim_param());
+    }
+  }
+  return SwapGatherNDWithInputNode(graph, gathernd_node, gathernd_input_node, 0);
+}
+
+static std::unordered_map<std::string, Handler> handlers = {
+    {"Add", AddHandler},
+    {"Gelu", SimpleHandler},
+    {"LayerNormalization", SimpleHandler},
+    {"MatMul", MatMulHandler},
+};
+
+static Status Delegate(std::string op_type, Graph& graph, Node* gathernd_node, Node* gathernd_input_node) {
+  if (handlers.count(op_type)) {
+    return handlers[op_type](graph, gathernd_node, gathernd_input_node);
+  } else {
+    return common::Status(common::ONNXRUNTIME, common::NOT_IMPLEMENTED, op_type + " handler is not implemented");
+  }
+}
+
+Status ComputationReductionTransformer::ApplyImpl(Graph& graph, bool& modified, int graph_level,
+                                                  const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& order = graph_viewer.GetNodesInTopologicalOrder();
+
+  for (auto index : order) {
+    auto* node_ptr = graph.GetNode(index);
+    if (!node_ptr)
+      // node was removed, this should not happen since we are not removing nodes.
+      continue;
+
+    auto& node = *node_ptr;
+    ORT_RETURN_IF_ERROR(Recurse(node, modified, graph_level, logger));
+
+    if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "GatherND", {1}, kOnnxDomain) ||
```

Comment:
Could a similar optimization apply to "Gather" ops as well? Or (probably with more differences) to "GatherElements", "Slice", "Split", etc.?

Reply:
Good point, thanks! In theory, yes, while it might take some time to cover all the cases since we don't have very good use cases in real models. I added a todo in the comment.  Let's refine GatherND cases well before making it more generic (if needed).